pkg/backup/restic.go

```diff
@@ -2,129 +2,45 @@ package backup
-	rt "k8s.io/apimachinery/pkg/runtime"
-	"k8s.io/apimachinery/pkg/util/runtime"
-	"k8s.io/apimachinery/pkg/watch"
-	"k8s.io/client-go/tools/cache"
-	"k8s.io/client-go/util/workqueue"
 )
 
 func (c *Controller) initResticWatcher() {
 	// TODO: Watch one Restic object, when support for Kubernetes 1.8 is dropped.
 	// ref: https://github.com/kubernetes/kubernetes/pull/53345
-	lw := &cache.ListWatch{
-		ListFunc: func(options metav1.ListOptions) (rt.Object, error) {
-			return c.stashClient.Restics(c.opt.Namespace).List(options)
-		},
-		WatchFunc: func(options metav1.ListOptions) (watch.Interface, error) {
-			return c.stashClient.Restics(c.opt.Namespace).Watch(options)
-		},
-	}
-
-	// create the workqueue
-	c.rQueue = workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "restic")
-
-	// Bind the workqueue to a cache with the help of an informer. This way we make sure that
-	// whenever the cache is updated, the pod key is added to the workqueue.
-	// Note that when we finally process the item from the workqueue, we might see a newer version
-	// of the Restic than the version which was responsible for triggering the update.
-	c.rIndexer, c.rInformer = cache.NewIndexerInformer(lw, &api.Restic{}, c.opt.ResyncPeriod, cache.ResourceEventHandlerFuncs{
-		AddFunc: func(obj interface{}) {
-			if r, ok := obj.(*api.Restic); ok && r.Name == c.opt.ResticName && r.IsValid() == nil {
-				key, err := cache.MetaNamespaceKeyFunc(obj)
-				if err == nil {
-					c.rQueue.Add(key)
-				}
-			}
-		},
-		UpdateFunc: func(old interface{}, new interface{}) {
-			oldObj, ok := old.(*api.Restic)
-			if !ok {
-				log.Errorln("Invalid Restic object")
-				return
-			}
-			newObj, ok := new.(*api.Restic)
-			if !ok {
-				log.Errorln("Invalid Restic object")
-				return
-			}
-			if !util.ResticEqual(oldObj, newObj) && newObj.Name == c.opt.ResticName && newObj.IsValid() == nil {
-				key, err := cache.MetaNamespaceKeyFunc(new)
-				if err == nil {
-					c.rQueue.Add(key)
-				}
-			}
-		},
-		DeleteFunc: func(obj interface{}) {
-			// IndexerInformer uses a delta queue, therefore for deletes we have to use this
-			// key function.
-			if r, ok := obj.(*api.Restic); ok && r.Name == c.opt.ResticName {
-				key, err := cache.DeletionHandlingMetaNamespaceKeyFunc(obj)
-				if err == nil {
-					c.rQueue.Add(key)
-				}
-			}
-		},
-	}, cache.Indexers{cache.NamespaceIndex: cache.MetaNamespaceIndexFunc})
-	c.rLister = stash_listers.NewResticLister(c.rIndexer)
-}
 
-func (c *Controller) runResticWatcher() {
-	for c.processNextRestic() {
-	}
-}
+	c.rInformer = c.stashInformerFactory.Stash().V1alpha1().Restics().Informer()
+	c.rQueue = queue.New("Restic", c.opt.MaxNumRequeues, c.opt.NumThreads, c.runResticScheduler)
+	c.rInformer.AddEventHandler(queue.NewEventHandler(c.rQueue.GetQueue(), func(oldObj, newObj interface{}) bool {
+		oldRestic, ok := oldObj.(*api.Restic)
+		if !ok {
+			log.Errorln("Invalid Restic object")
+			return false
```

Comment:
If old is invalid but new is valid than we should handle it ?

Reply:
fixed